FFmpegInterop/UncompressedAudioSampleProvider.cpp

```diff
@@ -17,161 +17,152 @@
+	tuple<IBuffer, int64_t, int64_t, std::map<GUID, IInspectable>> UncompressedAudioSampleProvider::GetSampleData()
 	{
-		av_frame_free(&m_pAvFrame);
-	}
-
-	// Free 
-	swr_free(&m_pSwrCtx);
-}
+		// Decode samples until we reach the minimum sample duration threshold or EOS
+		IBuffer sampleBuf{ nullptr };
+		int64_t pts{ -1 };
+		int64_t dur{ 0 };
+		bool firstDecodedSample{ true };
+		vector<uint8_t> compactedSampleBuf;
+
+		while (true)
+		{
+			// Get the next decoded sample
+			AVFrame_ptr frame;
+			try
+			{
+				frame = GetFrame();
+			}
+			catch (...)
+			{
+				const hresult hr{ to_hresult() };
+				if (hr != MF_E_END_OF_STREAM)
+				{
+					// Unexpected error. Rethrow.
+					throw;
+				}
+				else if (firstDecodedSample)
+				{
+					// No decoded samples. Rethrow.
+					throw;
+				}
+				else
+				{
+					// Return the decoded sample data we have
+					sampleBuf = make<FFmpegInteropBuffer>(move(compactedSampleBuf));
+					break;
+				}
+			}
 
-HRESULT UncompressedAudioSampleProvider::WriteAVPacketToStream(DataWriter^ dataWriter, AVPacket* avPacket)
-{
-	// Because each packet can contain multiple frames, we have already written the packet to the stream
-	// during the decode stage.
-	return S_OK;
-}
+			IBuffer curSampleBuf{ nullptr };
+			int curSampleCount{ 0 };
 
-HRESULT UncompressedAudioSampleProvider::ProcessDecodedFrame(DataWriter^ dataWriter)
-{
-	// Resample uncompressed frame to AV_SAMPLE_FMT_S16 PCM format that is expected by Media Element
-	uint8_t *resampledData = nullptr;
-	unsigned int aBufferSize = av_samples_alloc(&resampledData, NULL, m_pAvFrame->channels, m_pAvFrame->nb_samples, AV_SAMPLE_FMT_S16, 0);
-	int resampledDataSize = swr_convert(m_pSwrCtx, &resampledData, aBufferSize, (const uint8_t **)m_pAvFrame->extended_data, m_pAvFrame->nb_samples);
-	auto aBuffer = ref new Platform::Array<uint8_t>(resampledData, min(aBufferSize, (unsigned int)(resampledDataSize * m_pAvFrame->channels * av_get_bytes_per_sample(AV_SAMPLE_FMT_S16))));
-	dataWriter->WriteBytes(aBuffer);
-	av_freep(&resampledData);
-	av_frame_unref(m_pAvFrame);
-	av_frame_free(&m_pAvFrame);
-
-	return S_OK;
-}
+			if (m_swrContext == nullptr)
+			{
+				// Uncompressed frame is already in the desired output format
+				curSampleBuf = make<FFmpegInteropBuffer>(frame->buf[0]);
+				curSampleCount = frame->nb_samples;
+			}
+			else
+			{
+				// Resample uncompressed frame to desired output format
+				uint8_t* buf{ nullptr };
+				int bufSize{ av_samples_alloc(&buf, nullptr, frame->channels, frame->nb_samples, AV_SAMPLE_FMT_S16, 0) };
+				THROW_HR_IF_FFMPEG_FAILED(bufSize);
 
-MediaStreamSample^ UncompressedAudioSampleProvider::GetNextSample()
-{
-	// Similar to GetNextSample in MediaSampleProvider, 
-	// but we concatenate samples until reaching a minimum duration
-	DebugMessage(L"GetNextSample\n");
+				int resampledSamplesCount{ swr_convert(m_swrContext.get(), &buf, bufSize, const_cast<const uint8_t**>(frame->extended_data), frame->nb_samples) };
```

Comment:
>bufSize [](start = 69, length = 7)

This existed prior to your refactor, but per the documentation for swr_convert, we should be passing in the number of samples per channel in the output buffer here, instead of the output buffer size in bytes.